src/wayland/seat/touch.rs

```diff
@@ -0,0 +1,184 @@
+pub struct TouchHandle {
+    inner: Rc<RefCell<TouchInternal>>,
+}
+
+impl TouchHandle {
+    pub(crate) fn new() -> Self {
+        Self {
+            inner: Default::default(),
+        }
+    }
+
+    /// Register a new touch handle to this handler
+    ///
+    /// This should be done first, before anything else is done with this touch handle.
+    pub(crate) fn new_touch(&self, touch: WlTouch) {
+        self.inner.borrow_mut().known_handles.push(touch);
+    }
+
+    /// Notify clients about new touch points.
+    pub fn down(
+        &mut self,
+        time: u32,
+        surface: &WlSurface,
+        surface_offset: Point<i32, Logical>,
+        slot: TouchSlot,
+        location: Point<f64, Logical>,
+    ) {
+        self.inner
+            .borrow_mut()
+            .down(time, surface, surface_offset, slot, location);
+    }
+
+    /// Notify clients about touch point removal.
+    pub fn up(&self, time: u32, slot: TouchSlot) {
+        self.inner.borrow_mut().up(time, slot);
+    }
+
+    /// Notify clients about touch motion.
+    pub fn motion(&self, time: u32, slot: TouchSlot, location: Point<f64, Logical>) {
+        self.inner.borrow_mut().motion(time, slot, location);
+    }
+
+    /// Notify clients about touch shape changes.
+    pub fn shape(&self, slot: TouchSlot, major: f64, minor: f64) {
+        self.inner.borrow_mut().shape(slot, major, minor);
+    }
+
+    /// Notify clients about touch shape orientation.
+    pub fn orientation(&self, slot: TouchSlot, orientation: f64) {
+        self.inner.borrow_mut().orientation(slot, orientation);
+    }
+
+    /// Notify clients about touch cancellation.
+    ///
+    /// This should be sent by the compositor when the currently active touch
+    /// slot was recognized as a gesture.
+    pub fn cancel(&self) {
+        self.inner.borrow_mut().cancel();
+    }
+}
+
+/// Touch-slot focused Wayland client state.
+#[derive(Default, Debug)]
+struct TouchFocus {
+    surface_offset: Point<f64, Logical>,
+    handles: Vec<WlTouch>,
+}
+
+#[derive(Default, Debug)]
+struct TouchInternal {
+    known_handles: Vec<WlTouch>,
+    focus: HashMap<TouchSlot, TouchFocus>,
+}
+
+impl TouchInternal {
+    fn down(
+        &mut self,
+        time: u32,
+        surface: &WlSurface,
+        surface_offset: Point<i32, Logical>,
+        slot: TouchSlot,
+        location: Point<f64, Logical>,
+    ) {
+        // Update focused client state.
+        let focus = self.focus.entry(slot).or_default();
+        focus.surface_offset = surface_offset.to_f64();
+        focus.handles.clear();
+
+        // Select all WlTouch instances associated to the active WlSurface.
+        for handle in &self.known_handles {
+            if handle.as_ref().same_client_as(surface.as_ref()) {
+                focus.handles.push(handle.clone());
+            }
+        }
+
+        let (x, y) = (location - focus.surface_offset).into();
+        let serial = SERIAL_COUNTER.next_serial().into();
```

Comment:
The serial can be used by the client to request some action, so the compositor needs to be aware of them. Otherwise it can't implement it's own serial checking rules. 
So we should not hide any serials from the user.

@vberger Correct me if I'm wrong

Reply:
Just for my personal information: What kind of action can clients request here?

Reply:
I'm not familiar with touch based clients, so I don't know what they do often, but it can basically be anything, lets say xdg-shell move/resize, or show_window_menu for example.